mailbox/plugin/quota-mailing/src/test/java/org/apache/james/eventsourcing/EventSourcingSystemTest.java

```diff
@@ -0,0 +1,194 @@
+        EventSourcingSystem eventSourcingSystem = new EventSourcingSystem(
+            ImmutableSet.of(wordCuttingDispatcher(eventStore)),
+            ImmutableSet.of(subscriber),
+            eventStore);
+
+        eventSourcingSystem.dispatch(new MyCommand("This is a test"));
+
+        assertThat(subscriber.getData()).containsExactly("This", "is", "a", "test");
+    }
+
+    @Test
+    default void unknownCommandsShouldBeIgnored(EventStore eventStore) {
+        DataCollectorSubscriber subscriber = new DataCollectorSubscriber();
+        EventSourcingSystem eventSourcingSystem = new EventSourcingSystem(
+            ImmutableSet.of(wordCuttingDispatcher(eventStore)),
+            ImmutableSet.of(subscriber),
+            eventStore);
+
+        eventSourcingSystem.dispatch(new CommandDispatcher.Command() {});
+
+        assertThat(subscriber.getData()).isEmpty();
+    }
+
+    default CommandDispatcher.CommandHandler<MyCommand> simpleDispatcher(EventStore eventStore) {
+        return new CommandDispatcher.CommandHandler<MyCommand>() {
+            @Override
+            public Class<MyCommand> getCommandClass() {
+                return MyCommand.class;
+            }
+
+            @Override
+            public List<? extends Event> handle(MyCommand myCommand) {
+                EventStore.History history = eventStore.getEventsOfAggregate(AGGREGATE_ID);
+
+                return ImmutableList.of(new TestEvent(
+                    history.getNextEventId(),
+                    AGGREGATE_ID,
+                    myCommand.getPayload()));
+            }
+        };
+    }
+
+    default CommandDispatcher.CommandHandler<MyCommand> wordCuttingDispatcher(EventStore eventStore) {
+        return new CommandDispatcher.CommandHandler<MyCommand>() {
+            @Override
+            public Class<MyCommand> getCommandClass() {
+                return MyCommand.class;
+            }
+
+            @Override
+            public List<? extends Event> handle(MyCommand myCommand) {
+                EventStore.History history = eventStore.getEventsOfAggregate(AGGREGATE_ID);
+
+                EventIdIncrementer eventIdIncrementer = new EventIdIncrementer(history.getNextEventId());
+
+                return Splitter.on(" ")
+                    .splitToList(myCommand.getPayload())
+                    .stream()
+                    .map(word -> new TestEvent(
+                        eventIdIncrementer.next(),
+                        AGGREGATE_ID,
+                        word))
+                    .collect(Guavate.toImmutableList());
+            }
+        };
+    }
+
+    class EventIdIncrementer {
+        private EventId currentEventId;
+
+        public EventIdIncrementer(EventId base) {
+            this.currentEventId = base;
+        }
+
+        public EventId next() {
+            currentEventId = currentEventId.next();
+            return currentEventId;
+        }
+    }
+
+}
```

Comment:
we should check that when storage fail, subscribers are never called
we should also check what happen when a subscriber throws : do others subscribers get called ?

Reply:
Good test ideas, let's add them.